src/member.ts

```diff
@@ -0,0 +1,358 @@
+      shouldAccept: allowAll,
+    }
+  ) {
+    super();
+    this.id = id || makeID();
+    this.ownFeed = [];
+    this.ownIndex = 0;
+    this.knownFeeds = {
+      [this.id]: this.ownFeed,
+    };
+    this.memberIndexes = {
+      [this.id]: 0,
+    };
+    this.requests = {};
+    this.knownMembers = [initiator || this.id];
+    this.shouldAccept = shouldAccept;
+  }
+
+  sync(member: Member) {
+    // Iterate through member's knownFeeds
+    // Get any new data you don't have
+    // Get new data from members to knownFeeds
+    for (const id of Object.keys(member.knownFeeds)) {
+      const feed = member.knownFeeds[id];
+      const ownCopy = this.getFeedFor(id);
+      if (ownCopy.length < feed.length) {
+        this.updateFeedFor(id, feed);
+      }
+    }
+  }
+
+  processFeeds(): boolean {
+    let hasProcessed = false;
+    let hasResponded = false;
+    for (const id of Object.keys(this.knownFeeds)) {
+      const index = this.getMemberIndexFor(id);
+      const feed = this.getFeedFor(id);
+      const item = feed[index];
+
+      if (!item) continue;
+      hasProcessed = true;
+
+      if (isRequest(item)) {
+        this.trackRequest(item);
+
+        // TODO: Detect detect concurrent requests and merge
+        if (this.isMember()) {
+          hasResponded = true;
+          const shouldAccept = this.shouldAccept(item);
+          if (shouldAccept) this.acceptRequest(item);
+          else this.denyRequest(item);
+        }
+        this.incrementMemberIndexFor(id);
+      } else if (isResponse(item)) {
+        if (this.hasRequest(item.id)) {
+          const req = this.getRequest(item.id);
+          req.addResponse(id, item);
+          let neededSignatures = this.knownMembers.length;
+          if (req.operation === "remove") neededSignatures--;
+          const maxDenied = this.knownMembers.length - neededSignatures;
+          const state = req.getState(neededSignatures, maxDenied);
+          if (req.getState(neededSignatures) === "ready") {
+            if (req.operation === "add") {
+              if (isWho(req.req)) {
+                const who = req.req.who;
+                this.knownMembers.push(who);
+              }
+            } else if (req.operation === "remove") {
+              if (isWho(req.req)) {
+                const who = req.req.who;
+                this.knownMembers = this.knownMembers.filter(
+                  (id) => id !== who
+                );
+              }
+            }
+            req.finish();
+            if (this.isMember()) {
+              this.emit("block", req);
+            }
+          }
+          this.incrementMemberIndexFor(id);
+        } else {
+          continue;
+        }
+      } else {
+        console.warn("Got invalid item", item);
+        this.incrementMemberIndexFor(id);
+      }
+    }
+
+    if (hasProcessed) {
+      const otherRun = this.processFeeds()
+      return hasProcessed || otherRun;
```

Comment:
Ugh, endless recursion possible, right? 😅

Reply:
Err, what's the condition that causes it?

Reply:
Nothing at the moment, but my eyes look with stress 💦 on the code and I would be constantly worried if it introduces endless recursions.

Reply:
Is there a different approach I should take here? The current structure recurses until all the feeds have been processed (more specifically, until no items are processable).